src/parser/parser.yy

```diff
@@ -730,6 +738,37 @@ when_then_list
     }
     ;
 
+
+list_comprehension_expression
+    : L_BRACKET expression KW_IN expression KW_WHERE expression R_BRACKET {
+        if ($2->kind() != Expression::Kind::kLabel) {
+            throw nebula::GraphParser::syntax_error(@2, "The loop variable must be a label in list comprehension");
+        }
+        auto &innerVar = *(static_cast<const LabelExpression *>($2)->name());
+        auto *expr = new ListComprehensionExpression(new std::string(innerVar), $4, $6, nullptr);
+        nebula::graph::ParserUtil::rewriteLC(qctx, expr, innerVar);
+        $$ = expr;
+    }
+    | L_BRACKET expression KW_IN expression PIPE expression R_BRACKET {
+        if ($2->kind() != Expression::Kind::kLabel) {
+            throw nebula::GraphParser::syntax_error(@2, "The loop variable must be a label in list comprehension");
+        }
+        auto &innerVar = *(static_cast<const LabelExpression *>($2)->name());
+        auto *expr = new ListComprehensionExpression(new std::string(innerVar), $4, nullptr, $6);
+        nebula::graph::ParserUtil::rewriteLC(qctx, expr, innerVar);
+        $$ = expr;
+    }
+    | L_BRACKET expression KW_IN expression KW_WHERE expression PIPE expression R_BRACKET {
```

Comment:
Maybe you could define a `list_comprehension_generator_expr` for `expression KW_IN expression` syntax, and check the Expression Kind and return ListComprehensionExpression pointer. In `list_comprehension_expression`, you only need to set the mapper and filter and rewirte above expression.

Reply:
List_comprehension_generator_expr is not general. 
Because other expressions like `reduce` and `predicate` also has syntax likes `expression KW_IN expression`. They can't reuse `list_comprehension_generator_expr`